## Imports

In [1]:
import os
from collections import defaultdict

import pandas as pd
import numpy as np
from tensorflow import keras

2024-05-05 17:45:24.979907: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


## 1. Load Data

Last week we took a look through all our raw data

To allow for testing further down the line I've held out some data in advance that we'll be using later in the course. As such I've made a folder in the course project folder system called `train_test_data`

For this workshop please download:
- `train_test_data` folder and put within `data/`

Key for data:
- train_genes = gene_match_data for train samples
- test_genes = gene_match_data for test samples
- y_train = array of S/R target values
- y_train_ids = array of genome_ids in order of y_train
- y_test_ids = array of genome_ids in order of y_test

In [2]:
seed = 130

def load_data():
    """
    Load the data needed for Workshop 4
    """
    train_genes = pd.read_csv('../data/train_test_data/train_genes.csv')
    train_genes['genome_id'] = train_genes.genome_id.astype(str)
    test_genes = pd.read_csv('../data/train_test_data/test_genes.csv')
    test_genes['genome_id'] = test_genes.genome_id.astype(str)
    y_train = np.load('../data/train_test_data/y_train.npy', allow_pickle=True)
    y_train_ids = np.load('../data/train_test_data/train_ids.npy', allow_pickle=True).astype(str)
    y_test_ids = np.load('../data/train_test_data/test_ids.npy', allow_pickle=True).astype(str)

    return train_genes, test_genes, y_train, y_train_ids, y_test_ids

train_genes, test_genes, y_train, y_train_ids, y_test_ids = load_data()

In [3]:
train_genes.head(5)

genome_id              contig  \
0  562.22444  562.22444.con.0012   
1  562.22444  562.22444.con.0019   
2  562.22444  562.22444.con.0030   
3  562.22444  562.22444.con.0088   
4  562.22444  562.22444.con.0088   

                                            res_gene  match_start  match_end  \
0  gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol...        41274      42123   
1  gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ec...        60711      63126   
2  gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ec...        21837      25866   
3  gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFT...            0          3   
4  gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ec...            0          4   

         match_qual                                          query_str  \
0              849M  TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...   
1             2415M  ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...   
2             4029M  TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...   
3  3M135I1023M19I5M  GTGTCTAAAGAAAAATTTGAACGTACAAAACCGCACGTTAACGTTG...   
4  4M180I1022M19I5M  GTGCTCTCTCCTGAAGGGGAGAGCACTATAGTAAGGAATATAGCCG...   

                                        ref_gene_str  
0  TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...  
1  ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...  
2  TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...  
3  GcaTTCGACCAGATCGATAACGCGCCGGAAGAAAAAGCTCGTGGTA...  
4  GcatTCGACCAGATCGATAACGCGCCGGAAGAAAAAGCTCGTGGTA...

In [4]:
y_train[0:5], y_train_ids[0:5]

(array([['S'],
        ['S'],
        ['S'],
        ['S'],
        ['S']], dtype=object),
 array(['562.22444', '562.42705', '562.42733', '562.42728', '562.45644'],
       dtype='<U9'))

## 1. Presence / Absence Features

Our first, most simple feature set will be the presence/absence of each gene we've seen from our CARD alignment data.

In order to build these features we'll need to:
1. Find all unique res_genes, count which samples they're present/absent in
2. Look for correlations between genes
3. Remove highly correlated features (as seen in the assignment)

### 1a. Create a presence absence matrix for each sample and each gene

- We're going to leverage some pandas magic to make this really simple
- The logic can be quite complex manually
  - For each gene, search each samples
  - Store a list of 0/1 for each samples for each gene
  - Ensuring correct ordering
- Pandas can do this for us using `pivot_table`

In [5]:
def build_gene_presence_absence(dataset, ids, gene_names=None):
    """
    Build a matrix of samples to genes with 1 for present and 0 for absent

    Args:
      - dataset (pd.DataFrame): dataset of gene alignments
      - ids (list): ordering for IDs
      - gene_names (list): ordered list of unique genes (optional)
    """
    # If not providing genes take all unique genes from the data
    if gene_names is None:
        gene_names = dataset.res_gene.unique()

    # Count for all genes found within the data
    genes_counts = (
        dataset
        [['genome_id', 'res_gene']]
        .assign(count = 1)
        .pivot_table('count', index='genome_id', columns='res_gene')
        .fillna(0)
    )

    # Add genes missing from the data
    missing_genes = set(gene_names).difference(set(genes_counts.columns))
    if len(missing_genes) > 0:
        for gene in missing_genes:
            genes_counts[gene] = 0

    # Make sure to return in the same order as gene_names and the same sample order
    return genes_counts.loc[ids, gene_names]

In [6]:
train_presence_absence_df = build_gene_presence_absence(train_genes, y_train_ids)
print('\nShape of PA df:', train_presence_absence_df.shape, '\n')
train_presence_absence_df.head()


Shape of PA df: (748, 1428) 



res_gene   gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF  \
genome_id                                                            
562.22444                                                1.0         
562.42705                                                1.0         
562.42733                                                1.0         
562.42728                                                1.0         
562.45644                                                1.0         

res_gene   gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                1.0           
562.42733                                                1.0           
562.42728                                                1.0           
562.45644                                                1.0           

res_gene   gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                1.0           
562.42733                                                1.0           
562.42728                                                1.0           
562.45644                                                1.0           

res_gene   gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR  \
genome_id                                                        
562.22444                                                1.0     
562.42705                                                0.0     
562.42733                                                0.0     
562.42728                                                1.0     
562.45644                                                0.0     

res_gene   gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                0.0           
562.42733                                                0.0           
562.42728                                                1.0           
562.45644                                                0.0           

res_gene   gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO  \
genome_id                                                            
562.22444                                                1.0         
562.42705                                                1.0         
562.42733                                                1.0         
562.42728                                                1.0         
562.45644                                                1.0         

res_gene   gb|AP009048.1|-|3172159-3174052|ARO:3003316|Ecol_parE_FLO  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                1.0           
562.42733                                                1.0           
562.42728                                                1.0           
562.45644                                                1.0           

res_gene   gb|U00096.3|-|3470144-3471329|ARO:3003370|Ecol_EFTu_ENC  \
genome_id                                                            
562.22444                                                1.0         
562.42705                                                0.0         
562.42733                                                0.0         
562.42728                                                1.0         
562.45644                                                0.0         

res_gene   gb|AE006468.2|-|4038867-4041282|A

In [7]:
genes_of_interest = [
    'gb|AAAGNS010000063.1|-|144-1005|ARO:3000966|TEM-103',
    'gb|AB023477.1|+|0-861|ARO:3001082|SHV-24',
    'Random Missing Gene',
    'gb|AB089595.1|+|0-1206|ARO:3000166|tet(B)',
]
build_gene_presence_absence(train_genes, y_train_ids, genes_of_interest).head()

res_gene   gb|AAAGNS010000063.1|-|144-1005|ARO:3000966|TEM-103  \
genome_id                                                        
562.22444                                                0.0     
562.42705                                                1.0     
562.42733                                                1.0     
562.42728                                                0.0     
562.45644                                                1.0     

res_gene   gb|AB023477.1|+|0-861|ARO:3001082|SHV-24  Random Missing Gene  \
genome_id                                                                  
562.22444                                       0.0                    0   
562.42705                                       0.0                    0   
562.42733                                       0.0                    0   
562.42728                                       0.0                    0   
562.45644                                       0.0                    0   

res_gene   gb|AB089595.1|+|0-1206|ARO:3000166|tet(B)  
genome_id                                             
562.22444                                        0.0  
562.42705                                        0.0  
562.42733                                        0.0  
562.42728                                        1.0  
562.45644                                        1.0

### 2b. Review Correlations and Remove Identical Features


If we leave these, they can cause issues during modeling (unidentifiability/incorrect feature importances and inferences).

Two options:
1. Calculate correlations and cluster
2. Look at identical presence/absence and select one

We'll go for the optiob 2 but you can also try clustering in your project if you believe it will help model performance.

Again this seems somewhat complex - we need to look across all samples and check for identical 1/0 arrays. BUT:

- Pandas to the rescue once again

In [8]:
# Transpose the data so that genes are our rows
train_presence_absence_transposed = train_presence_absence_df.transpose()

# Use "drop_duplicates()" to remove identical rows (will just keep the first)
train_unique_presence_absence = train_presence_absence_transposed.drop_duplicates()

# Transpose back so genes are columns
train_presence_absence_df = train_unique_presence_absence.transpose()

In [9]:
print('\nShape of PA df:', train_presence_absence_df.shape, '\n')
train_presence_absence_df.head()


Shape of PA df: (748, 159) 



res_gene   gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF  \
genome_id                                                            
562.22444                                                1.0         
562.42705                                                1.0         
562.42733                                                1.0         
562.42728                                                1.0         
562.45644                                                1.0         

res_gene   gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                1.0           
562.42733                                                1.0           
562.42728                                                1.0           
562.45644                                                1.0           

res_gene   gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                1.0           
562.42733                                                1.0           
562.42728                                                1.0           
562.45644                                                1.0           

res_gene   gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR  \
genome_id                                                        
562.22444                                                1.0     
562.42705                                                0.0     
562.42733                                                0.0     
562.42728                                                1.0     
562.45644                                                0.0     

res_gene   gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                0.0           
562.42733                                                0.0           
562.42728                                                1.0           
562.45644                                                0.0           

res_gene   gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO  \
genome_id                                                            
562.22444                                                1.0         
562.42705                                                1.0         
562.42733                                                1.0         
562.42728                                                1.0         
562.45644                                                1.0         

res_gene   gb|AE006468.2|-|4038867-4041282|ARO:3003307|Sser_gyrB_FLO  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                0.0           
562.42733                                                0.0           
562.42728                                                1.0           
562.45644                                                1.0           

res_gene   gb|HG738867.1|+|2934266-2935658|ARO:3003890|Ecol_UhpT_FOF  \
genome_id                                                              
562.22444                                                1.0           
562.42705                                                1.0           
562.42733                                                1.0           
562.42728                                                1.0           
562.45644                                                1.0           

res_gene   gb|HG738867.1|-|223

### 2c. Also then need to subset our test data

Always need to make sure our test data is in the same shape and format

In [10]:
# Get just the unique 159 genes from our training data
unique_train_genes = list(train_presence_absence_df.columns)
test_presence_absence_df = build_gene_presence_absence(test_genes, y_test_ids, unique_train_genes)

In [11]:
test_presence_absence_df.head()

res_gene   gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF  \
genome_id                                                            
562.42833                                                1.0         
562.42739                                                1.0         
562.22823                                                1.0         
562.45646                                                1.0         
562.22547                                                1.0         

res_gene   gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU  \
genome_id                                                              
562.42833                                                1.0           
562.42739                                                1.0           
562.22823                                                1.0           
562.45646                                                1.0           
562.22547                                                1.0           

res_gene   gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF  \
genome_id                                                              
562.42833                                                1.0           
562.42739                                                1.0           
562.22823                                                1.0           
562.45646                                                1.0           
562.22547                                                1.0           

res_gene   gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR  \
genome_id                                                        
562.42833                                                0.0     
562.42739                                                0.0     
562.22823                                                1.0     
562.45646                                                0.0     
562.22547                                                1.0     

res_gene   gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV  \
genome_id                                                              
562.42833                                                0.0           
562.42739                                                0.0           
562.22823                                                1.0           
562.45646                                                0.0           
562.22547                                                1.0           

res_gene   gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO  \
genome_id                                                            
562.42833                                                1.0         
562.42739                                                1.0         
562.22823                                                1.0         
562.45646                                                1.0         
562.22547                                                1.0         

res_gene   gb|AE006468.2|-|4038867-4041282|ARO:3003307|Sser_gyrB_FLO  \
genome_id                                                              
562.42833                                                0.0           
562.42739                                                0.0           
562.22823                                                1.0           
562.45646                                                0.0           
562.22547                                                0.0           

res_gene   gb|HG738867.1|+|2934266-2935658|ARO:3003890|Ecol_UhpT_FOF  \
genome_id                                                              
562.42833                                                1.0           
562.42739                                                1.0           
562.22823                                                1.0           
562.45646                                                1.0           
562.22547                                                1.0           

res_gene   gb|HG738867.1|-|223

## 3. Kmer Features

Kmers are a representation of the raw sequencing data.

K being a parameter setting the length of the sequence (e.g. 2-mer or 5-mer)

There are two main options for kmerizing data:
1. Utilize Python to count the sequence data
2. Utilize a command line tool (see [Jellyfish](https://github.com/zippav/Jellyfish-2))

In this tutorial we'll learn how to generate kmers manually on a small subset of data.

Kmer counting is an expensive operation, for using Kmer features in your project - I've already generated an output file across all the data using Jellyfish (we'll loop back to this next week)


1. A simple for loop will work perfectly for this
2. We can select each chunk of 7 nucleotides at each position
3. We'll need to figure out how to keep track of this efficiently

In [12]:
def count_kmers(sequence, k=7):
    """
    For a single sequence, slide over all nucleotides and count each k chunk

    Args:
      - sequence (str): raw nucleotide data
      - k (int): size of kmers
    """
    kmer_counts = defaultdict(int)
    for i_start in range(len(sequence) - (k-1)):
        kmer = sequence[i_start:i_start + k]
        kmer_counts[kmer] += 1
    
    return kmer_counts

In [13]:
# Test with a sequence
test_sequence = 'ACGTGTGTAAGACGTGTGGCGA'
count_kmers('ACGTGTGTAAGACGTGTGGCGA')

defaultdict(int,
            {'ACGTGTG': 2,
             'CGTGTGT': 1,
             'GTGTGTA': 1,
             'TGTGTAA': 1,
             'GTGTAAG': 1,
             'TGTAAGA': 1,
             'GTAAGAC': 1,
             'TAAGACG': 1,
             'AAGACGT': 1,
             'AGACGTG': 1,
             'GACGTGT': 1,
             'CGTGTGG': 1,
             'GTGTGGC': 1,
             'TGTGGCG': 1,
             'GTGGCGA': 1})

### 3b. Apply this approach for each sample

We have a function which can turn a sequence into counts of kmers

This still isn't a usable "feature" for modeling though, we now need to apply across our training data, lets take a look at the data again:

In [14]:
train_genes.head(3)

genome_id              contig  \
0  562.22444  562.22444.con.0012   
1  562.22444  562.22444.con.0019   
2  562.22444  562.22444.con.0030   

                                            res_gene  match_start  match_end  \
0  gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol...        41274      42123   
1  gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ec...        60711      63126   
2  gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ec...        21837      25866   

  match_qual                                          query_str  \
0       849M  TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...   
1      2415M  ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...   
2      4029M  TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...   

                                        ref_gene_str  
0  TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...  
1  ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...  
2  TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...

Our data isn't in a single sequence per sample, it's per sample per gene.

We can either:
1. Count kmers per gene sequence and then aggregate per sample
2. Aggregate sequences to a single sequence per sample then count

We're going to take option 2 as it makes tracking the kmers much easier

To make this computationally feasible we're going to try on just the first 5 samples

In [15]:
def count_kmers_per_sample(dataset, sample_id_col='genome_id', gene_col='ref_gene_str'):

    # Agg to full gene string per sample
    seq_per_sample = dataset.groupby(sample_id_col).agg({'ref_gene_str': ''.join}).reset_index()
    seq_per_sample['ref_gene_str'] = seq_per_sample['ref_gene_str'].str.upper()

    kmer_counts = {}
    for ref_name, sequence in seq_per_sample[[sample_id_col, gene_col]].to_records(index=False):
        kmer_counts[ref_name] = count_kmers(sequence)

    # Convert to DataFrame, fill empty with zero and transpose
    kmer_counts_df = pd.DataFrame(kmer_counts).fillna(0).transpose()
    
    return kmer_counts_df

In [16]:
first_5_samples = train_genes.genome_id.unique()[0:5]
train_genes_first_five = train_genes[train_genes.genome_id.isin(first_5_samples)]
test_kmer_counts = count_kmers_per_sample(train_genes_first_five)

In [17]:
test_kmer_counts.head()

TTACTCA  TACTCAT  ACTCATA  CTCATAG  TCATAGC  CATAGCG  ATAGCGT  \
562.22444     21.0      8.0      2.0      2.0      6.0      7.0     16.0   
562.42705     19.0    200.0     18.0      3.0      8.0     10.0     15.0   
562.42728      3.0      1.0      1.0      3.0      5.0      3.0     13.0   
562.42733     10.0    199.0     47.0      5.0      9.0      6.0     15.0   
562.45644      5.0      2.0      2.0      3.0      7.0     15.0     18.0   

           TAGCGTT  AGCGTTT  GCGTTTG  ...  TTCTTAG  ATGCTAG  CTAGCAC  TAGCACT  \
562.22444      9.0     20.0     15.0  ...      0.0      0.0      0.0      0.0   
562.42705      5.0    234.0     21.0  ...      0.0      0.0      0.0      0.0   
562.42728     15.0     26.0     18.0  ...      0.0      0.0      0.0      0.0   
562.42733     10.0    224.0     20.0  ...      0.0      0.0      0.0      0.0   
562.45644     26.0     46.0     19.0  ...      1.0      1.0      1.0      1.0   

           TCTATGG  TTAGACC  GGACAAG  GACCCTA  TCCTATT  TTGGACC  
562.22444      0.0      0.0      0.0      0.0      0.0      0.0  
562.42705      0.0      0.0      0.0      0.0      0.0      0.0  
562.42728      0.0      0.0      0.0      0.0      0.0      0.0  
562.42733      0.0      0.0      0.0      0.0      0.0      0.0  
562.45644      1.0      1.0      1.0      1.0      1.0      1.0  

[5 rows x 16245 columns]

In [18]:
# Check unique possible 7-mers
4**7

16384

### Final Note

We've counted every kmer that was present in each sample

BUT: we haven't account for the kmers which weren't seen anywhere in the samples
- 16,245 kmers present in the first 5 genomes
- We need the feature matrix to be consistent for future unseen
- Therefore need to account for ALL possible kmers

Our feature matrix would need to be `N x 16384`

This could get very large if we use a high K! (Remember 10-mers had over 1 million unique features)
